iup/__init__.py

```diff
@@ -0,0 +1,739 @@
+import polars as pl
+import datetime as dt
+from sklearn.linear_model import LinearRegression
+import numpy as np
+import abc
+from typing_extensions import Self
+from typing import List
+
+
+class UptakeData(pl.DataFrame, metaclass=abc.ABCMeta):
+    """
+    Abstract class for different forms of uptake data.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initialize an UptakeData object as a polars data frame plus validation.
+        """
+        super().__init__(*args, **kwargs)
+        self.validate()
+
+    def validate(self):
+        """
+        Validate that an UptakeData object has the three key columns:
+        date, grouping factors, and uptake estimate (% of population).
+        """
+        self.assert_columns_found(["date", "region", "estimate"])
+        self.assert_columns_type(["date"], pl.Date)
+        self.assert_columns_type(["region"], pl.Utf8)
+        self.assert_columns_type(["estimate"], pl.Float64)
+
+    def assert_columns_found(self, columns: List[str]):
+        """
+        Verify that expected columns are found.
+
+        Parameters
+        columns: List[str]
+            names of expected columns
+        """
+        for col in columns:
+            assert col in self.columns, f"Column {col} is expected but not found."
+
+    def assert_columns_type(self, columns: List[str], dtype):
+        """
+        Verify that columns have the expected type.
+
+        Parameters
+        columns: List[str]
+            names of columns for which to check type
+        dtype:
+            data type for each listed column
+        """
+        for col in columns:
+            assert (
+                self[col].dtype == dtype
+            ), f"Column {col} should be {dtype} but is {self[col].dtype}"
+
+    def with_columns(self, *args, **kwargs):
```

Comment:
This is an interesting point, that any polars verb will return a normal `pl.DataFrame`, rather than an instance of this object

Personally, I think that's OK, and the user could wrap `UptakeData(df)` any time they want to validate or assert that `df` is actually uptake data

Because otherwise we'd need to figure out some way to ensure that `.rename()`, `.filter()`, etc. all do the same thing

Reply:
Issue #42